3p/iframe-transport-client.js

```diff
@@ -33,61 +33,134 @@ export class IframeTransportClient {
+                  'Received malformed event in ' + this.win_.location.href);
+              this.contextFor_(event.creativeId).dispatch(event.message);
             } catch (e) {
               user().error(TAG_,
-                  'Exception in callback passed to onAnalyticsEvent: ' +
-                  e.message);
+                  'Exception in callback passed to onAnalyticsEvent',
+                  e);
             }
           });
         });
   }
 
   /**
+   * Retrieves/creates a context object to pass events pertaining to a
+   * particular creative.
+   * @param {string} creativeId The ID of the creative
+   * @returns {!IframeTransportContext}
+   * @private
+   */
+  contextFor_(creativeId) {
+    return this.creativeIdToContext_[creativeId] ||
+        (this.creativeIdToContext_[creativeId] =
+            new IframeTransportContext(this.win_, this.iframeMessagingClient_,
+            creativeId, this.vendor_));
+  }
+
+  /**
+   * Gets the IframeMessagingClient.
+   * @returns {!IframeMessagingClient}
+   * @VisibleForTesting
+   */
+  getIframeMessagingClient() {
+    return this.iframeMessagingClient_;
+  }
+}
+
+/**
+ * A context object to be passed along with event data.
+ */
+class IframeTransportContext {
+  /**
+   * @param {!Window} win
+   * @param {!IframeMessagingClient} iframeMessagingClient
+   * @param {string} creativeId The ID of the creative that the event
+   *     pertains to.
+   * @param {string} vendor The 3p vendor name
+   */
+  constructor(win, iframeMessagingClient, creativeId, vendor) {
+    /** @private {!IframeMessagingClient} */
+    this.iframeMessagingClient_ = iframeMessagingClient;
+
+    /** @private @const {!Object} */
+    this.baseMessage_ = {creativeId, vendor};
+
+    /** @private {?function(string)} */
+    this.listener_ = null;
+
+    user().assert(win['onNewAmpAnalyticsInstance'] &&
+        typeof win['onNewAmpAnalyticsInstance'] == 'function',
+        'Must implement onNewAmpAnalyticsInstance in ' + win.location.href);
+    win['onNewAmpAnalyticsInstance'](this);
+  }
+
+  /**
    * Registers a callback function to be called when an AMP analytics event
    * is received.
    * Note that calling this a second time will result in the first listener
    * being removed - the events will not be sent to both callbacks.
-   * @param {function(string,string)} callback
+   * @param {!function(string)} listener
    */
-  onAnalyticsEvent(callback) {
-    this.listener_ = callback;
+  onAnalyticsEvent(listener) {
+    this.listener_ = listener;
   }
 
   /**
-   * Gets the IframeMessagingClient
-   * @returns {!IframeMessagingClient}
-   * @VisibleForTesting
+   * Receives an event from IframeTransportClient, and passes it along to
+   * the creative that this context represents.
+   * @param {string} event
+   */
+  dispatch(event) {
+    this.listener_ && this.listener_(event);
+  }
+
+  /**
+   * Sends a response message back to the creative.
+   * @param {!Object<string,string>} message
```

Comment:
nit: space after ,
message sounds like a string. what about `data`

Reply:
I was calling it "response" yesterday, but @keithwrightbos said to change it to "message". I will change it to "data".